codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1047 @@
+            public void close(ChannelPromise promise) {
+                if (closePromise.isDone()) {
+                    promise.setFailure(new ClosedChannelException());
+                    return;
+                }
+                if (inClose) {
+                    return;
+                }
+                inClose = true;
+                try {
+                    closePending = false;
+                    fireChannelReadPending = false;
+
+                    // Only ever send a reset frame if the connection is still alive as otherwise it makes no sense at
+                    // all anyway.
+                    if (parent().isActive() && !streamClosedWithoutError && isStreamIdValid(stream().id())) {
+                        Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).stream(stream());
+                        write(resetFrame, unsafe().voidPromise());
+                        flush();
+                    }
+
+                    if (inboundBuffer != null) {
+                        for (;;) {
+                            Object msg = inboundBuffer.poll();
+                            if (msg == null) {
+                                break;
+                            }
+                            ReferenceCountUtil.release(msg);
+                        }
+                    }
+
+                    pipeline().fireChannelInactive();
+                    if (isRegistered()) {
+                        deregister(unsafe().voidPromise());
+                    }
+                    promise.setSuccess();
+                    closePromise.setSuccess();
+                } finally {
+                    inClose = false;
+                }
+            }
+
+            @Override
+            public void closeForcibly() {
+                close(unsafe().voidPromise());
+            }
+
+            @Override
+            public void deregister(ChannelPromise promise) {
+                if (registered) {
+                    registered = true;
+                    promise.setSuccess();
+                    pipeline().fireChannelUnregistered();
+                } else {
+                    promise.setFailure(new IllegalStateException("Not registered"));
+                }
+            }
+
+            @Override
+            public void beginRead() {
+                if (readInProgress || !isActive()) {
+                    return;
+                }
+
+                final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+                allocHandle.reset(config());
+                if (inboundBuffer == null) {
+                    if (closePending) {
+                        unsafe.closeForcibly();
+                    } else {
+                        readInProgress = true;
+                    }
+                    return;
+                }
+
+                do {
+                    Object m = inboundBuffer.poll();
+                    if (m == null && closePending) {
```

Comment:
can you add a comment explaining why `m == null && closePending` are checked together here and `null` is allowed to propagate to `doRead0`. The following would be more intuitive:

```java
                    if (m == null) {
                        if (closePending) {
                            pipeline().fireChannelReadComplete();
                            unsafe.closeForcibly();
                        }
                        return;
                    }
```